winforms/src/toga_winforms/window.py

```diff
@@ -137,13 +137,27 @@ def winforms_FormClosing(self, sender, event):
             if not self.interface.closeable:
                 # Closeability is implemented by shortcutting the close handler.
                 event.Cancel = True
+
             elif self.interface.on_close._raw:
                 # If there is an on_close event handler, process it;
                 # but then cancel the close event. If the result of
                 # on_close handling indicates the window should close,
                 # then it will be manually triggered as part of that
                 # result handling.
                 self.interface.on_close(self)
+                # Close the app when the last window closes.
+                if len(self.interface.app.windows) == 0:
+                    if self.interface.app.on_exit:
+                        self.interface.app.on_exit(self.interface.app)
+                event.Cancel = True
+
+            # Close the window when no on_close handler is provided.
```

Comment:
Replicating logic like this should be a red flag. If the close logic is *that* similar, it should be refactored. 

That said, it's not 100% clear to me that this change is correct. What are the conditions that necessitate this change?

Reply:
I originally missed the examples you had in the PR description; I'm still not entirely convinced this is correct. Mind you, I'm not entirely convinced the *original* is correct either; closing the "last" window shouldn't be the trigger for on_exit logic - it should be the close of the Main Window, whether it's the last window or not. I think we might have been leaning a bit on the fact that the lifecycle of the MainForm was previously coupled to the event loop. If this change has broken the assumption that make that work, we should be looking to migrate this on_close handling to be main-window specific.

Reply:
I mean previously, the app exited when the `main_window` closed, but since this decoupling led to defining of `main_window` as an optional. Hence I made it so that the app closes when the last window closes, or else the app will continue to run. But I agree we should modify the closing behavior. We also need to discuss about when should the app exit, now that `main_window` is an optional.